etna/libs/ts2vec/ts2vec.py

```diff
@@ -0,0 +1,387 @@
+                    x = x[:, window_offset: window_offset + self.max_train_length]
+                x = x.to(self.device)
+
+                ts_l = x.size(1)
+                crop_l = np.random.randint(low=2 ** (self.temporal_unit + 1), high=ts_l + 1)
+                crop_left = np.random.randint(ts_l - crop_l + 1)
+                crop_right = crop_left + crop_l
+                crop_eleft = np.random.randint(crop_left + 1)
+                crop_eright = np.random.randint(low=crop_right, high=ts_l + 1)
+                crop_offset = np.random.randint(low=-crop_eleft, high=ts_l - crop_eright + 1, size=x.size(0))
+
+                optimizer.zero_grad()
+
+                out1 = self._net(take_per_row(x, crop_offset + crop_eleft, crop_right - crop_eleft))
+                out1 = out1[:, -crop_l:]
+
+                out2 = self._net(take_per_row(x, crop_offset + crop_left, crop_eright - crop_left))
+                out2 = out2[:, :crop_l]
+
+                loss = hierarchical_contrastive_loss(
+                    out1,
+                    out2,
+                    temporal_unit=self.temporal_unit
+                )
+
+                loss.backward()
+                optimizer.step()
+                self.net.update_parameters(self._net)
+
+                cum_loss += loss.item()
+                n_epoch_iters += 1
+
+                self.n_iters += 1
+
+                if self.after_iter_callback is not None:
+                    self.after_iter_callback(self, loss.item())
+
+            if interrupted:
+                break
+
+            cum_loss /= n_epoch_iters
+            loss_log.append(cum_loss)
+            if verbose:
+                print(f"Epoch #{self.n_epochs}: loss={cum_loss}")
+            self.n_epochs += 1
+
+            if self.after_epoch_callback is not None:
+                self.after_epoch_callback(self, cum_loss)
+
+        return loss_log
+
+    def _eval_with_pooling_segment(self, x, mask=None, slicing=None):
+        out = self.net(x.to(self.device, non_blocking=True), mask)
+        if slicing is not None:
+            out = out[:, slicing]
+        out = F.max_pool1d(
+            out.transpose(1, 2),
+            kernel_size=out.size(1),
+        ).transpose(1, 2)
+
+        return out.cpu()
+
+    def _eval_with_pooling_window(self, x, mask=None, slicing=None, encoding_window=None):
+        out = self.net(x.to(self.device, non_blocking=True), mask)
+        if encoding_window is not None:
+            out = F.max_pool1d(
+                out.transpose(1, 2),
+                kernel_size=encoding_window,
+                stride=1,
+                padding=encoding_window // 2
+            ).transpose(1, 2)
+            if encoding_window % 2 == 0:
+                out = out[:, :-1]
+        if slicing is not None:
+            out = out[:, slicing]
+
+        return out.cpu()
+
+    def encode_segment(self, data, mask=None, causal=False, sliding_length=None, sliding_padding=0, batch_size=None):
```

Comment:
Divided one method into two modes

Reply:
If you have already decided to separate methods, then why is there a `causal` parameter here? For segment encoding it is always False

Reply:
Its almost the same as encode_window. Why we split it?